service-base/src/main/java/org/eclipse/hono/service/resourcelimits/PrometheusBasedResourceLimitChecks.java

```diff
@@ -627,76 +647,146 @@ private Long extractLongValue(final JsonObject response, final Span span) {
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     * @throws IllegalArgumentException if target date-time is before effective since date-time.
      */
     long calculateEffectiveLimit(
-            final OffsetDateTime effectiveSince,
-            final OffsetDateTime targetDateTime,
-            final PeriodMode mode,
+            final Instant effectiveSince,
+            final Instant targetDateTime,
+            final PeriodMode periodType,
             final long configuredLimit) {
-        if (PeriodMode.MONTHLY.equals(mode)
-                && configuredLimit > 0
-                && !targetDateTime.isBefore(effectiveSince)
-                && YearMonth.from(targetDateTime).equals(YearMonth.from(effectiveSince))
-                && effectiveSince.getDayOfMonth() != 1) {
-            final OffsetDateTime lastDayOfMonth = effectiveSince.with(TemporalAdjusters.lastDayOfMonth());
-            final long daysBetween = ChronoUnit.DAYS
-                    .between(effectiveSince, lastDayOfMonth) + 1;
-            return Double.valueOf(Math.ceil(daysBetween * configuredLimit / lastDayOfMonth.getDayOfMonth()))
-                    .longValue();
+
+        Objects.requireNonNull(effectiveSince, "effective since");
+        Objects.requireNonNull(targetDateTime, "target date-time");
+        Objects.requireNonNull(periodType, "period mode");
+
+        if (targetDateTime.isBefore(effectiveSince)) {
+            throw new IllegalArgumentException("target date-time must not be before effective since");
+        }
+
+        // we only need to calculate the effective limit if we are in the initial accounting period
+        // of a monthly plan
+        if (PeriodMode.MONTHLY.equals(periodType) && configuredLimit > 0) {
+
+            final ZonedDateTime effectiveSinceZonedDateTime = ZonedDateTime.ofInstant(effectiveSince, ZoneOffset.UTC);
+            final ZonedDateTime targetZonedDateTime = ZonedDateTime.ofInstant(targetDateTime, ZoneOffset.UTC);
+
+            if (YearMonth.from(targetZonedDateTime).equals(YearMonth.from(effectiveSinceZonedDateTime))) {
+
+                final ZonedDateTime startOfNextAccountingPeriod = effectiveSinceZonedDateTime
+                        .with(TemporalAdjusters.firstDayOfNextMonth())
+                        .withHour(0)
+                        .withMinute(0)
+                        .withSecond(0)
+                        .withNano(0);
+                final long minutesTillStartOfNextAccountingPeriod = Math.max(1, Duration
+                        .between(effectiveSinceZonedDateTime, startOfNextAccountingPeriod)
+                        .toMinutes());
+                final long lengthOfCurrentMonthInMinutes = 60 * 24 * effectiveSinceZonedDateTime
+                        .range(ChronoField.DAY_OF_MONTH).getMaximum();
+                return Double.valueOf(Math.ceil(minutesTillStartOfNextAccountingPeriod * configuredLimit / lengthOfCurrentMonthInMinutes))
```

Comment:
The result won't be the rounded up division result here because the division is done in an integer context and so the division result is truncated before it gets used in `Math.ceil()`.

Reply:
true, I have changed that